doc/guides/smf/index.rst

```diff
@@ -0,0 +1,267 @@
+	struct s_object {
+		/* This must be first */
+		struct smf_ctx ctx;
+
+		/* Other state specific data add here */
+	} s_obj;
+
+	/* State S0 */
+	static void s0_entry(void *o) {}
+	static void s0_run(void *o)
+	{
+		set_state(SMF_CTX(s_obj), &demo_states[S1]);
+	}
+	static void s0_exit(void *o) {}
+
+	/* State S1 */
+	static void s1_entry(void *o) {}
+	static void s1_run(void *o)
+	{
+		set_state(SMF_CTX(s_obj), &demo_states[S2]);
+	}
+	static void s1_exit(void *o) {}
+
+	/* State S2 */
+	static void s2_entry(void *o) {}
+	static void s2_run(void *o)
+	{
+		set_state(SMF_CTX(s_obj), &demo_states[S0]);
+	}
+	static void s2_exit(void *o) {}
+
+	/* Populate state table */
+	static const struct smf_state demo_states[] {
+		[S0] = FSM_CREATE_STATE(s0_entry, s0_run, s0_exit),
+		[S1] = FSM_CREATE_STATE(s1_entry, s1_run, s1_exit),
+		[S2] = FSM_CREATE_STATE(s2_entry, s2_run, s2_exit),
+	};
+
+	void main(void)
+	{
+		/* Set initial state */
+		set_stat(SMF_CTX(s_obj), &demo_states[S0]);
+
+		/* Run the state machine */
+		while(1) {
+			run_state(SMF_CTX(s_obj));
+			k_msleep(1000);
+		}
+	}
+
+Hierarchical State Machine Example
+==================================
+This example turns the following state diagram into code using the SMF, where
+S0 and S1 share a parent state and S0 is the initial state.
+
+.. figure:: img/hierarchical.png
+    :align: center
+    :alt: Hierarchial state machine diagram
+
+Code::
+
+	#include <smf.h>
+
+	/* Forward declaration of state table */
+	static const struct smf_state demo_states[];
+
+	/* List of demo states */
+	enum demo_state { PARENT, S0, S1, S2 };
+
+	/* User defined object */
+	struct s_object {
+		/* This must be first */
+		struct smf_ctx ctx;
+
+		/* Other state specific data add here */
+	} s_obj;
+
+	/* Parent State */
+	static void parent_entry(void *o) {}
+	static void parent_run(void *o) {}
+	static void parent_exit(void *o) {}
+
+	/* State S0 */
+	static void s0_entry(void *o) {}
+	static void s0_run(void *o)
+	{
+		set_state(SMF_CTX(s_obj), &demo_states[S1]);
+	}
+	static void s0_exit(void *o) {}
+
+	/* State S1 */
+	static void s1_entry(void *o) {}
+	static void s1_run(void *o)
+	{
+		set_state(SMF_CTX(s_obj), &demo_states[S2]);
+	}
+	static void s1_exit(void *o) {}
+
+	/* State S2 */
+	static void s2_entry(void *o) {}
+	static void s2_run(void *o)
+	{
+		set_state(SMF_CTX(s_obj), &demo_states[S0]);
+	}
+	static void s2_exit(void *o) {}
+
+	/* Populate state table */
+	static const struct smf_state demo_states[] {
+		[PARENT] = HSM_CREATE_STATE(parent_entry, parent_run, parent_exit, NULL);
+		[S0] = HSM_CREATE_STATE(s0_entry, s0_run, s0_exit, &demo_states[PARENT]),
+		[S1] = HSM_CREATE_STATE(s1_entry, s1_run, s1_exit, &demo_states[PARENT]),
+		[S2] = HSM_CREATE_STATE(s2_entry, s2_run, s2_exit),
+	};
+
+	void main(void)
+	{
+		/* Set initial state */
+		set_stat(SMF_CTX(s_obj), &demo_states[S0]);
+
+		/* Run the state machine */
+		while(1) {
+			run_state(SMF_CTX(s_obj));
```

Comment:
should error handling be allowed?

Reply:
I added a way to return an error and terminate the state machine